### PR TITLE
Mobile messaging

### LINKED ISSUE #34544: Add mobile messaging with deep links
Internal description of the feature: pe5pgL-qo-p2 

### Description

We'd like to add mobile-specific messaging for orders-related emails to promote mobile apps. Those messages will contain deep links, which will navigate users to a specific area in the mobile app (or to the landing page, where users can learn more about apps).

For now, we're focusing on a new order email and manage order/accept payment kind of messages.

### PR DESCRIPTION
**Pull request by wzieba**
### All Submissions:

-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #34544.
Project thread (internal): pe5pgL-qo-p2

This PR adds marketing messages to "New Order" mails. The following algorithm selects the messages: 

<details>
<summary>Old copy selection logic (iteration 2)</summary>

</details>
Up-to-date copy selection logic (iteration 3)
image

#### TC1: Store and order eligible for IPP
1. Assert that store is eligible for IPP (is based in the US with USD currency or in CA with CAD)
2. Create a new, IPP-eligible Order (empty order in Processing state is okay. All requirements: PdfdoF-V-p2)
3. Assert that received "New Order" mail has the "Collect payment easily (...)" message in the footer.

#### TC2: Order not eligible for IPP
1. With the same site as above, create a new non-eligible IPP Order (e.g. set payment method Paypal or add subscription products)
2. Assert that received "New Order" mail has the "Manage the order (...)" message in the footer.

#### TC3: Store not eligible for IPP
1. Change the store currency to other than USD (if the store was the US) or CAD (if the store was the CA).
2. Create a new, IPP-eligible Order
3. Assert that received "New Order" mail has the "Manage the order (...)" message in the footer.

#### TC4: No Jetpack
1. Revert the store's currency change from TC3 if needed.
2. Disable Jetpack
3. Create a new, IPP-eligible Order
4. Assert that received the "Process your orders on the go (...) message in the footer.

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+		?int $blog_id,
+		DateTime $now
+	): ?string {
+		try {
+			$last_mobile_used = self::get_closer_mobile_usage_date();
+
+			$used_app_in_last_month = $last_mobile_used->diff( $now )->days <= self::OPEN_ORDER_INTERVAL_DAYS;
+			$has_jetpack            = $blog_id !== null;
+
+			if ( $used_app_in_last_month && $has_jetpack ) {
```
Comment (wzieba) [inline plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: This logic block is WIP, see (internal):pe5pgL-EE-p2

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+	 *
+	 * @param WC_Order $order order that the conditions are checked for.
+	 *
+	 * @return true if order is eligible, false otherwise
+	 */
+	private static function is_order_in_person_payment_eligible( WC_Order $order ): bool {
+		return in_array( $order->get_status(), array( 'pending', 'on-hold', 'processing' ), true ) &&
+			   in_array( $order->get_payment_method(), array( 'cod', 'woocommerce_payments', 'none' ), true ) && // phpcs:ignore WordPress.WhiteSpace.PrecisionAlignment.Found
+			   self::is_store_in_person_payment_eligible() && // phpcs:ignore WordPress.WhiteSpace.PrecisionAlignment.Found
+			   ! $order->is_paid() && // phpcs:ignore WordPress.WhiteSpace.PrecisionAlignment.Found
```
Comment (wzieba) [inline plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: I have no idea how to fix this: neither `phpcbf` nor PHPStorm WordPress code style doesn't fix those issues. I also couldn't fix this manually (or maybe I don't understand the message).

File: plugins/woocommerce/templates/emails/email-mobile-messaging.php [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * HOWEVER, on occasion WooCommerce will need to update template files and you
+ * (the theme developer) will need to copy the new files to your theme to
+ * maintain compatibility. We try to do this as little as possible, but it does
+ * happen. When this occurs the version of the template file will be bumped and
+ * the readme will list any important changes.
+ *
+ * @see https://docs.woocommerce.com/document/template-structure/
+ * @package WooCommerce\Templates\Emails
+ * @version 6.9.0
```
Comment (wzieba) [inline plugins/woocommerce/templates/emails/email-mobile-messaging.php]: I'll change it to 7.0 or 7.1.

File: plugins/woocommerce/tests/php/includes/class-wc-mobile-messaging-handler-test.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+/**
+ * Class WC_Mobile_Messaging_Handler_Test file.
+ *
+ * @package WooCommerce\Tests
+ */
+class WC_Mobile_Messaging_Handler_Test extends WC_Unit_Test_Case {
```
Comment (wzieba) [inline plugins/woocommerce/tests/php/includes/class-wc-mobile-messaging-handler-test.php]: I'll add more unit tests later, ideally to fully cover `WC_Mobile_Messaging_Handler` - I didn't do it yet as the SUT class can heavily change at this point.

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+	/**
+	 * Prepares mobile messaging with a deep link
+	 *
+	 * @param WC_Order $order order that mobile message is created for.
+	 * @param ?int     $blog_id  of blog to make a deep link for.
+	 * @param DateTime $now      current DateTime.
+	 *
+	 * @return string|null
+	 */
+	public static function prepare_mobile_message(
```
Comment (wzieba) [inline plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: I have concerns about "staticness" of this class. Is it acceptable? Or should I (somehow) inject `WCTracker` and make it not static? Or it's not needed?

File: plugins/woocommerce/changelog/Mobile deep link email footer [hunk 6]
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: enhancement
+
+Adds mobile deep link email footer
```
Comment (wzieba) [inline plugins/woocommerce/changelog/Mobile deep link email footer]: - [x] To change later.

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+	}
+
+	/**
+	 * Returns if order is eligible to accept In-Person Payments
+	 *
+	 * @param WC_Order $order order that the conditions are checked for.
+	 *
+	 * @return true if order is eligible, false otherwise
+	 */
+	private static function is_order_in_person_payment_eligible( WC_Order $order ): bool {
```
Comment (wzieba) [inline plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: Logic was taken from internal documentation: PdfdoF-V-p2 except for one condition

- [x] TODO: add "Order does NOT have any subscription products" condition

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+		return in_array( $order->get_status(), array( 'pending', 'on-hold', 'processing' ), true ) &&
+			   in_array( $order->get_payment_method(), array( 'cod', 'woocommerce_payments', 'none' ), true ) && // phpcs:ignore WordPress.WhiteSpace.PrecisionAlignment.Found
+			   self::is_store_in_person_payment_eligible() && // phpcs:ignore WordPress.WhiteSpace.PrecisionAlignment.Found
```
Comment (wzieba) [inline plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: This condition is duplicated (at this point we already checked if store is IPP eligible) but I find this check needed to make sure that `is_order_in_person_payment_eligible` function does exactly what it says it does - checks all conditions.

I believe it's so minor that not even measurable in terms of performance so decided to leave it.

**Review by wzieba — COMMENTED** [review]
Hi @coreymckrill 👋 

I've moved to the original repository and added the improvements you mentioned in the first review - I hope I understood them correctly.

Also, I've added a few lines of new code as the requirements changed.

Please feel free to take a look at the code and share if things go in the right direction. Thanks!

Ps. I don't understand why tests are failing 🤔 they seem to run out of memory but I'm not sure if this is something I could affect with code changes in this PR.

File: plugins/woocommerce/includes/emails/class-wc-email.php [hunk 9]
```diff
@@ -256,6 +256,25 @@ public function __construct() {
 
 		add_action( 'phpmailer_init', array( $this, 'handle_multipart' ) );
 		add_action( 'woocommerce_update_options_email_' . $this->id, array( $this, 'process_admin_options' ) );
+		add_action( 'woocommerce_email_footer', array( $this, 'mobile_messaging' ), 9 ); // Run before the default email footer.
```
Comment (wzieba) [inline plugins/woocommerce/includes/emails/class-wc-email.php]: I'm not sure why there's `priority 9`. Should `mobile_messaging` be executed before `woocommerce_email_footer` (as your comment says) or I didn't understand it?

image

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php (hunk 2 above)
Comment (coreymckrill) [reply plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: I think it means that each condition is using individual space characters in order to line up with the condition on the line above it, but the code style rules say that all indenting should be done with tabs, not spaces.

Honestly, though, the logic is a little hard to follow with so many conditions. My suggestion would be to assign each of those conditions to a separate boolean variable and then just check that they're all true at the end. Something like this:

```php
$has_status = in_array( $order->get_status(), array( 'pending', 'on-hold', 'processing' ), true );
$has_method = in_array( $order->get_payment_method(), array( 'cod', 'woocommerce_payments', 'none' ), true );
$store_is_eligible = self::is_store_in_person_payment_eligible();
$order_is_not_paid = ! $order->is_paid();
$order_is_not_refunded = ! empty( $order->get_refunds() );

return $has_status && $has_method && $store_is_eligible && $order_is_not_paid && $order_is_not_refunded;
```

File: plugins/woocommerce/includes/emails/class-wc-email.php (hunk 9 above)
Comment (coreymckrill) [reply plugins/woocommerce/includes/emails/class-wc-email.php]: Yeah, lower priority number means it will happen sooner. The default priority is `10`. We want this message to be output before the footer.

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php (hunk 5 above)
Comment (coreymckrill) [reply plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: I think static methods are ok in this case. One thing that you're doing that isn't really used anywhere else in WooCommerce (as far as I know) is specifying the return type. While I think that's technically compatible with the minimum version of PHP required by WC, it's a lot stricter than usual. But we can check with someone on the core team about it.

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php (hunk 8 above)
Comment (coreymckrill) [reply plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: (See my other comment on the next line down)

File: plugins/woocommerce/templates/emails/email-mobile-messaging.php (hunk 3 above)
Comment (coreymckrill) [reply plugins/woocommerce/templates/emails/email-mobile-messaging.php]: Might be best to leave it as `x.x.x` or something for now, so we don't forget about it later.

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+			'https://woocommerce.com/mobile/payments'
+		);
+
+		return sprintf(
+			wp_kses_data(
+			/* translators: %s: Email link */
+				__(
+					'<a href="%1$s">Accept payments</a> with a card reader in our mobile app\.</br><a href="%2$s">Learn more about In-Person Payments\.</a>',
+					'woocommerce'
+				)
```
Comment (coreymckrill) [inline plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: 1. I would probably move the `translators` comment down to a line above the beginning of the string, to make sure it gets captured by the translation string parser. (Same with the other two instances farther down)
    ```
	__(
		/* translators: %s: Email link */
		'<a href="%1$s">Accept payments</a>...'
    ```
2. I don't think the period characters need to be escaped in this string.
3. `</br>` should be `<br />` instead.

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+			),
+			esc_url( $deep_link_url ),
+			esc_url( 'https://woocommerce.com/in-person-payments/' )
```
Comment (coreymckrill) [inline plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: This doesn't need the `esc_url` wrapper since it's a static string instead of a variable.

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php (hunk 10 above)
Comment (wzieba) [reply plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: > I would probably move the translators comment down to a line above the beginning of the string, to make sure it gets captured by the translation string parser.

I've tried doing this:

image

and phpcs returns:

```
| ERROR | A gettext call containing placeholders was found, but was not accompanied by a "translators:" comment on the line above to clarify the meaning of the placeholders. (WordPress.WP.I18n.MissingTranslatorsComment)
```

so I'm not sure we can do this.

**Comment by wzieba** [comment]
I added some improvements @coreymckrill. As PR starts to have a meaningful shape, I'll start working on unit tests.

----

I think that unit test jobs are canceled because performance tests for placing an order fail. It's probably expected - we started to do more operations to prepare a new order email. Should we increase the expected baseline times? Or should we approach it differently, WDYT?

**Comment by wzieba** [comment]
I've extracted some IPP-specific methods to a separate class to make it easier to unit-test and to reduce `WC_Mobile_Messaging_Handler` responsibilities.

~But I see I'm making some mistakes with the unit tests. It seems that those unit tests are not independent - `WC_Ipp_Functions_Test` is passing locally, but I see it's failing on CI (when all tests are executed). Do you have an idea or suggestion on how I could make those unit tests independent?~

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php (hunk 10 above)
Comment (coreymckrill) [reply plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: Oh interesting. I assumed the translators comment needed to be adjacent to the actual text string, but I guess it needs to be adjacent to the gettext function ( `__()` ) instead. So, you're already doing it right.

**Comment by coreymckrill** [comment]
@wzieba 

> I think that unit test jobs are canceled because performance tests for placing an order fail.

I'm not totally understanding the issue. Can you point me to the tests that are failing?

> It's probably expected - we started to do more operations to prepare a new order email. Should we increase the expected baseline times? Or should we approach it differently, WDYT?

If the failures are actually directly related to these changes then I think we should increase the baselines, but I'm not convinced yet that that's what's going on...

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php (hunk 5 above)
Comment (coreymckrill) [reply plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: It looks like there are a few other places in the codebase (all within the `src` directory) that use return types, so I guess this is fine.

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+				'order_id' => absint( $order_id ),
+			),
+			'https://woocommerce.com/mobile/payments'
+		);
+
+		return sprintf(
+			wp_kses_data(
+			/* translators: %s: Email link */
+				__(
+					'<a href="%1$s">Collect payments easily</a> from your customers anywhere with our mobile app.',
```
Comment (coreymckrill) [inline plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]:
```suggestion
					'<a href="%s">Collect payments easily</a> from your customers anywhere with our mobile app.',
```

The enumerated token syntax isn't needed since there's only one token in the string.

**Comment by wzieba** [comment]
> I'm not totally understanding the issue. Can you point me to the tests that are failing?

I think it happened e.g. [here](https://github.com/woocommerce/woocommerce/runs/8090449897?check_suite_focus=true#step:8:1192) but it seems it doesn't happen anymore.

File: plugins/woocommerce/includes/wc-ipp-functions.php [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+ * @package WooCommerce\Functions
+ * @version 7.0
+ */
+
+/**
+ * Returns if order is eligible to accept In-Person Payments
+ *
+ * @param WC_Order $order order that the conditions are checked for.
+ *
+ * @return true if order is eligible, false otherwise
```
Comment (coreymckrill) [inline plugins/woocommerce/includes/wc-ipp-functions.php]:
```suggestion
 * @return bool True if order is eligible, false otherwise.
```

The `@return` type should be shown as `bool`. Right now my IDE interprets this as the function always returning specifically `true`. Same goes for the other functions in this file.

File: plugins/woocommerce/tests/php/includes/class-wc-mobile-messaging-handler-test.php [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+	private static $initial_country = '';
+	/**
+	 * @var string $initial_currency that is set on site which is a platform for unit tests, before running tests in this test suite.
+	 */
+	private static $initial_currency = '';
+
+	/**
+	 * Saves values of initial country and currency before running test suite.
+	 */
+	public static function setUpBeforeClass(): void {
```
Comment (coreymckrill) [inline plugins/woocommerce/tests/php/includes/class-wc-mobile-messaging-handler-test.php]:
```suggestion
	public static function wpSetUpBeforeClass(): void {
```

Probably best to use the specific WP child method for this instead of the generic phpunit one.

File: plugins/woocommerce/tests/php/includes/class-wc-mobile-messaging-handler-test.php [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+		self::$initial_country  = WC()->countries->get_base_country();
+		self::$initial_currency = get_woocommerce_currency();
+	}
+
+	/**
+	 * Restores initial values of country and currency after running test suite.
+	 */
+	public static function tearDownAfterClass(): void {
```
Comment (coreymckrill) [inline plugins/woocommerce/tests/php/includes/class-wc-mobile-messaging-handler-test.php]:
```suggestion
	public static function wpTearDownAfterClass(): void {
```

Probably best to use the specific WP child method for this instead of the generic phpunit one.

File: plugins/woocommerce/tests/php/includes/class-wc-mobile-messaging-handler-test.php [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+	 */
+	public function test_tracker_reports_only_android_usage() {
+		$now = $this->prepare_timeline_with_valid_last_mobile_app_usage();
+
+		$mobile_message = WC_Mobile_Messaging_Handler::prepare_mobile_message( new WC_Order(), self::BLOG_ID, $now );
+
+		$this->assertEquals(
+			'<a href="https://woocommerce.com/mobile/order?blog_id=' . self::BLOG_ID . '&#038;order_id=' . 0 . '">Manage the order</a> with the app.',
+			$mobile_message
+		);
```
Comment (coreymckrill) [inline plugins/woocommerce/tests/php/includes/class-wc-mobile-messaging-handler-test.php]: Instead of asserting the exact string value, could this just test whether the return is a _string_ or _null_? That way the wording could change in the future without having to also update the unit test, while still being a useful test.

File: plugins/woocommerce/tests/php/includes/class-wc-mobile-messaging-handler-test.php [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+			'woocommerce_mobile_app_usage',
+			array()
+		);
+
+		$mobile_message = WC_Mobile_Messaging_Handler::prepare_mobile_message( new WC_Order(), self::BLOG_ID, $now );
+
+		$this->assertEquals(
+			'Process your orders on the go. <a href="https://woocommerce.com/mobile/">Get the app</a>. Powered by Jetpack.',
+			$mobile_message
+		);
```
Comment (coreymckrill) [inline plugins/woocommerce/tests/php/includes/class-wc-mobile-messaging-handler-test.php]: This one would be harder to test without specifying the exact wording. One idea might be to look for the substring `href="https://woocommerce.com/mobile/"` since, if I'm not mistaken, that particular URL is only used when no mobile usage is detected.

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+ *
+ * @package WooCommerce\Emails
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * WC_Mobile_Messaging_Handler
+ */
+class WC_Mobile_Messaging_Handler {
```
Comment (barryhughes) [inline plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: Thinking of [this guidance](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/README.md):

> Ideally, all the new code for WooCommerce should consist of classes following the PSR-4 naming and living in this directory, and the code in [the `includes` directory](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/includes/README.md) should receive the minimum amount of changes required for bug fixing.

What do you think of moving this to the `src/` directory and to the `Automattic\WooCommerce` namespace? Additionally, if we do not intend to expose this as part of our public API, we could place it within the `Internal` sub-namespace.

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+
+	const OPEN_ORDER_INTERVAL_DAYS = 30;
```
Comment (barryhughes) [inline plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: So far, we only use this from within this class. We could consider marking it as private (unless you anticipate referencing it from elsewhere in the near future, of course):

```suggestion
	private const OPEN_ORDER_INTERVAL_DAYS = 30;
```

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+
+	/**
+	 * Prepares mobile messaging with a deep link
+	 *
+	 * @param WC_Order $order order that mobile message is created for.
+	 * @param ?int     $blog_id  of blog to make a deep link for.
+	 * @param DateTime $now      current DateTime.
+	 *
+	 * @return string|null
```
Comment (barryhughes) [inline plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: For consistency, what do you think of either following this convention (to match what we use a few lines above, in the case of `@param ?int $blog_id` and in the method signature itself)?

```suggestion
	 * @return ?string
```

Or indeed the reverse: use the `string|null` and `int|null` "union type" style within docblocks? ... Not vital we change this but it strikes me as being more readable when we are consistent.

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+
+/**
+ * WC_Mobile_Messaging_Handler
+ */
+class WC_Mobile_Messaging_Handler {
+
+	const OPEN_ORDER_INTERVAL_DAYS = 30;
+
+	/**
+	 * Prepares mobile messaging with a deep link
```
Comment (barryhughes) [inline plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]:
```suggestion
	 * Prepares mobile messaging with a deep link.
```

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+	 *
+	 * @param WC_Order $order order that mobile message is created for.
+	 * @param ?int     $blog_id  of blog to make a deep link for.
```
Comment (barryhughes) [inline plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: I was initially just a little confused by [line 35](https://github.com/woocommerce/woocommerce/pull/34467/files#diff-198890d4431c1f21d2d0a9d3ee3b9d8846cd83ec80bf2b028032459158ff7b1aR35):

```
$has_jetpack            = null !== $blog_id;
```

As it made me wonder what the blog ID has to do with whether Jetpack is present. Now I've read `WC_Email::mobile_messaging()` it's a little clearer, but maybe we could also make a small amendment here:

```suggestion
	 * @param ?int     $blog_id  of blog to make a deep link for (will be null if Jetpack is not enabled).
```

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+	public static function accept_payment_message(): string {
+		return sprintf(
+			wp_kses_data(
+			/* translators: %s: Email link */
+				__(
+					'<a href="%s">Collect payments easily</a> from your customers anywhere with our mobile app.',
+					'woocommerce'
+				)
+			),
+			'https://woocommerce.com/mobile/payments/'
```
Comment (barryhughes) [inline plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: Something we could consider:

```suggestion
			esc_html(
				/* translators: 1: opening link tag 2: closing link tag. */
				__(
					'%1$sCollect payments easily%2$s from your customers anywhere with our mobile app.',
					'woocommerce'
				)
			),
			'<a href="https://woocommerce.com/mobile/payments/">',
			'</a>'
```

- `esc_html()` incurs less overhead, and is also stricter
- Moves the HTML out of the translatable string (better for the translator, safer for us)
- Fixes the indentation of the translator note

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+
+		return sprintf(
+			wp_kses_data(
+			/* translators: %s: Email link */
+				__(
+					'<a href="%s">Manage the order</a> with the app.',
+					'woocommerce'
+				)
+			),
+			esc_url( $deep_link_url )
```
Comment (barryhughes) [inline plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: Similar to an earlier note, we could consider:

```suggestion
			esc_html(
				/* translators: 1: opening link tag 2: closing link tag. */
				__(
					'%1$sManage the order%2$s with the app.',
					'woocommerce'
				)
			),
			'<a href="' . esc_url( $deep_link_url ) . '">',
			'</a>'
```

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+	public static function no_app_message(): string {
+		return sprintf(
+			wp_kses_data(
+			/* translators: %s: Email link */
+				__(
+					'Process your orders on the go. <a href="%s">Get the app</a>. Powered by Jetpack.',
+					'woocommerce'
+				)
+			),
+			'https://woocommerce.com/mobile/'
```
Comment (barryhughes) [inline plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: Similar to earlier notes, how about:

```suggestion
			esc_html(
				/* translators: 1: opening link tag 2: closing link tag. */
				__(
					'Process your orders on the go. %1$sGet the app%2$s. Powered by Jetpack.',
					'woocommerce'
				)
			),
			'<a href="https://woocommerce.com/mobile/">',
			'</a>'
```

File: plugins/woocommerce/includes/emails/class-wc-email.php [hunk 26]
```diff
@@ -256,6 +256,27 @@ public function __construct() {
...
+	}
+
+	/**
+	 * Add mobile messaging.
+	 */
+	public function mobile_messaging() {
+		if ( ! $this instanceof WC_Email_New_Order ) {
```
Comment (barryhughes) [inline plugins/woocommerce/includes/emails/class-wc-email.php]: Curious, why not move this method (and the matching `add_action()` call) to `WC_Email_New_Order`? 

Just thinking that this code—which is specific to that other class and is not applicable to all emails—would then be grouped with other code also specific to new order emails.

File: plugins/woocommerce/includes/wc-ipp-functions.php [hunk 27]
```diff
@@ -0,0 +1,57 @@
+<?php
```
Comment (barryhughes) [inline plugins/woocommerce/includes/wc-ipp-functions.php]: As earlier, we generally don't want to add new code to `includes/`. We can make exceptions of course: but do you think there might be a case here for setting this up as a set of static methods belonging to a new class in our `Automattic\WooCommerce` namespace/the `Internal` sub-namespace)?

Since that would also give us the advantage of autoloading, we could ditch [this line](https://github.com/woocommerce/woocommerce/pull/34467/files#diff-0095ea46d160c94a1276a0eed1d86a5dbc91387c39ac46aae6abd0a1d1d314b9R35).

File: plugins/woocommerce/includes/wc-ipp-functions.php [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+	$has_status            = in_array( $order->get_status(), array( 'pending', 'on-hold', 'processing' ), true );
+	$has_payment_method    = in_array( $order->get_payment_method(), array( 'cod', 'woocommerce_payments', 'none' ), true );
+	$order_is_not_paid     = null === $order->get_date_paid();
+	$order_is_not_refunded = empty( $order->get_refunds() );
+
+	$order_has_no_subscription_products = true;
+	foreach ( $order->get_items() as $item ) {
+		$product = $item->get_product();
+
+		if ( $product->is_type( 'subscription' ) ) {
```
Comment (barryhughes) [inline plugins/woocommerce/includes/wc-ipp-functions.php]: Something to consider:

```suggestion
		$product = $item->get_product();

		if ( is_object( $product ) && $product->is_type( 'subscription' ) ) {
```

Though it is presumably unusual, `get_items()` is documented as possibly returning `bool`.

**Review by barryhughes — CHANGES_REQUESTED** [review]

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php (hunk 18 above)
Comment (coreymckrill) [reply plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: Ah, I wasn't aware, or at least had forgotten about that guidance. I guess it makes sense for these functions since they're encapsulated as a class. 👍

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php (hunk 23 above)
Comment (coreymckrill) [reply plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: I was under the impression that this was harder for translators to deal with, but it there's already precedent elsewhere in the codebase, sounds good to me. If we go this route, it could be simplified further to:


```suggestion
			/* translators: 1: opening link tag 2: closing link tag. */
			esc_html__(
				'%1$sCollect payments easily%2$s from your customers anywhere with our mobile app.',
				'woocommerce'
			),
			'<a href="https://woocommerce.com/mobile/payments/">',
			'</a>'
```

Comment (barryhughes) [reply plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: There could well be differing opinions on the matter, though it's definitely not without precedent (one of many cases that can be found in GlotPress):

translating-links

I'd push for this strategy because:

- We can add classes, change the target attribute, etc, without altering the translatable string (and without having to wait on translators submitting updates).
- It lets us use stricter escaping, and so can help prevent malicious translation attacks.

> If we go this route, it could be simplified further to...

:100:

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php (hunk 20 above)
Comment (wzieba) [reply plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: Yes, I completely agree. I've decided to stick with `?string` as it's more compact.

File: plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php (hunk 22 above)
Comment (wzieba) [reply plugins/woocommerce/includes/class-wc-mobile-messaging-handler.php]: Yes, that was indeed confusing. Thanks, I've added the suggested clarification.

File: plugins/woocommerce/includes/wc-ipp-functions.php (hunk 28 above)
Comment (wzieba) [reply plugins/woocommerce/includes/wc-ipp-functions.php]: Just to better understand: if `get_product()` would return `bool`, the `is_type` executed on it would cause crash, right?

File: plugins/woocommerce/includes/emails/class-wc-email.php (hunk 26 above)
Comment (wzieba) [reply plugins/woocommerce/includes/emails/class-wc-email.php]: Initially, I wasn't sure if "New Order" was the only type of mail to which we wanted to add messages to. But as it's clear now, I'll move this method and `add_action` directly to `WC_Email_New_Order`, thanks!

**Comment by wzieba** [comment]
Thank you for the review @barryhughes, I learned a few things!

I think I've applied all suggestions, please ping me if I didn't or I understood you incorrectly.

After moving files, I broke the feature. When creating a new Order, I have this:
image

Do you maybe see in the code what I am missing? I thought that the import on `email-mobile-messaging.php#20` should be enough, but it seems it's not.

**Comment by wzieba** [comment]
In 5029618 I've added `blog_id` parameter to the deep link for tracking reasons (internal communication: p6JqRr-7ou-p2). cc: @toupper

**Comment by toupper** [comment]
Thanks for addressing the tracking as well @wzieba :) On p6JqRr-7ou-p2 we discuss about the best way to pass the values we want to track through the URL. Since we already track `utm_source` and `utm_content` (please note that they don't track `blog_id` but `blog_url` which is not related to our case here), I thought that we can use those so the alpha team doesn't have to change anything. In that case, the parameter name for the `blog_id` would be `utm_content`. e.g `woocommerce.com/mobile/orders/details?blog_id=123&order_id=345&utm_source=mobile_deeplink_orders&utm_content=123456 `being 123456 the blog id of the user. Please consider that when implementing the tracking parameters :)

**Comment by wzieba** [comment]
Thanks @toupper ! Sure, good idea - I'll update the code to meet those requirements.

So heads up @barryhughes and @coreymckrill - I'm now working on a few improvements. I'll re-request your review when it'll be ready. Thank you!

**Comment by barryhughes** [comment]
Sounds good, thanks for the work so far!

> After moving files, I broke the feature. When creating a new Order, I have this: [error]

Following up on this one, in the first instance it is probably worth running `composer dump` (if you haven't already tried doing so), to update the classmaps.

File: plugins/woocommerce/includes/wc-ipp-functions.php (hunk 28 above)
Comment (barryhughes) [reply plugins/woocommerce/includes/wc-ipp-functions.php]: Exactly ... we'd get a fatal error. 

I don't know for sure that there is a chain of events that would lead to us actually receiving `(bool) false` at this point—but the *potential* is there, so being defensive is something to consider.

**Comment by wzieba** [comment]
@toupper 👋
In d8cf172 I've added `utm_source` and `utm_content` query parameters. `utm_source` can be:
- `mobile_deeplink_payments`
- `mobile_deeplink_orders_details`
- `mobile_deeplink_no_app`

and `utm_content` will be `blog_id` (or `0` if site doesn't have Jetpack). I think it meets the requirements discussed in p6JqRr-7ou-p2, right?

----

@barryhughes @coreymckrill 👋

In addition to the above-described change, in 6ee6341 I've updated the logic for selecting the correct copy. I've also updated the diagram in the PR description to spot the difference quickly. Ready for another round of review!

> (...) in the first instance it is probably worth running composer dump

It worked like a charm, thanks!

**Review by coreymckrill — APPROVED** [review]
👍 Nice work getting this put together and including unit tests!
